prefetcher/va_ampm_lite/va_ampm_lite.cc

```diff
@@ -15,33 +14,31 @@ auto va_ampm_lite::page_and_offset(T addr) -> std::pair<champsim::page_number, b
 bool va_ampm_lite::check_cl_access(champsim::block_number v_addr)
 {
   auto [vpn, page_offset] = page_and_offset(v_addr);
-  auto region = std::find_if(std::begin(regions), std::end(regions), [vpn = vpn](auto x) { return x.vpn == vpn; });
+  auto region = regions.check_hit(region_type{vpn});
 
-  return (region != std::end(regions)) && region->access_map.at(page_offset.to<std::size_t>());
+  return (region.has_value() && region->access_map.at(page_offset.to<std::size_t>()));
 }
 
 bool va_ampm_lite::check_cl_prefetch(champsim::block_number v_addr)
 {
   auto [vpn, page_offset] = page_and_offset(v_addr);
-  auto region = std::find_if(std::begin(regions), std::end(regions), [vpn = vpn](auto x) { return x.vpn == vpn; });
+  auto region = regions.check_hit(region_type{vpn});
 
-  return (region != std::end(regions)) && region->prefetch_map.at(page_offset.to<std::size_t>());
+  return (region.has_value() && region->prefetch_map.at(page_offset.to<std::size_t>()));
 }
 
 uint32_t va_ampm_lite::prefetcher_cache_operate(champsim::address addr, champsim::address ip, uint8_t cache_hit, bool useful_prefetch, access_type type,
                                                 uint32_t metadata_in)
 {
   auto [current_vpn, page_offset] = page_and_offset(addr);
   champsim::block_number block_addr{addr};
-  auto demand_region = std::find_if(std::begin(regions), std::end(regions), [vpn = current_vpn](auto x) { return x.vpn == vpn; });
+  auto demand_region = regions.check_hit(region_type{current_vpn});
```

Comment:
I think the issue is probably here. This region is returned by value, so when it gets updated, only the copy is updated, so the AM/PM doesn't get written back.

Reply:
Okay. It feels bad, but I just call regions.fill() to write back the change.
I guess it makes sense that it doesn't pass by reference, since if you are going to modify a cache value it should be done as a fill.

I am running some test now. I will let you know if this fixes the problems.

Reply:
That's the standard usage. This is the same thing that would happen for a hardware array: it needs to be written back after being updated.